src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerConfigOptions.cs

```diff
@@ -19,5 +20,12 @@ public abstract class AnalyzerConfigOptions
         /// Get an analyzer config value for the given key, using the <see cref="KeyComparer"/>.
         /// </summary>
         public abstract bool TryGetValue(string key, [NotNullWhen(true)] out string? value);
+
+        /// <summary>
+        /// Enumerates keys of all available options in no specific order.
+        /// </summary>
+        /// <exception cref="NotImplementedException">Not implemented by the derived type.</exception>
+        public virtual IEnumerable<string> Keys
```

Comment:
My main concern is whether we're going to have perf concerns in a release or two using `IEnumerable<string>` instead of `ImmutableArray<string>` or similar. What are your thoughts on that?

Reply:
The implementation can (and does) return pre-allocated `IEnumerable` instance. The scenario where enumerating all keys is needed does a lot of parsing and allocations already. Allocating an enumerator from a pre-allocated IEnumerable instance is negligible. We can also always add more complex API whose default implementation is to call `Keys` but can be overridden to provide allocation-free enumeration.

Reply:
Honestly, I'm less interested in the allocations and more interested in the two virtual calls per iteration of enumeration, but I'm ok to go with this for now.